### PR TITLE
EOS-22465: Present any HW resource as FRU based on Storage Enclosure …

### PR DESCRIPTION
**Pull request by gargpalak048**
…representation

## Problem Statement:

Present any HW resource as FRU based on Storage Enclosure representation

## Solution Design:

Updated the Enclosure sensors/actuators to adopt the new resource type

## Coding:

* [ ] Did you follow coding standard and verified it with flake8? 

* [] Did your address all Codacy issues?

## Testing:

* [x] Unittest updated and all unittests are passing?

* [x] Sanity/Self tests are updated (if applicable)?
updated enclosure tests resource type hwerever applicable

* [x] Manual testing done covering happy path and non-happy path?

## Integration:

* [ ] If there is any interface change, did you communicate it to other components?
* [ ] If there is any interface change, other componenent gate gatekeepers are ready to accept the change?
* [ ] Did you complete deployment test if applicable?

## PR checklist:
* [x] Did you add Jira number to the PR?

* [x] DCO and cla-signed?

### CODE REVIEW

**Comment by cortx-admin** [comment]
Can one of the admins verify this patch?

File: low-level/framework/platforms/realstor/realstor_enclosure.py [hunk 1]
```diff
@@ -96,6 +96,13 @@ class RealStorEnclosure(StorageEnclosure):
     DATA_FORMAT_JSON = "json"
     FAULT_KEY = "unhealthy-component"
 
+    # Realstor FRUs mapping
+    fru_mapping = {
```
Comment (shriya-deshmukh) [inline low-level/framework/platforms/realstor/realstor_enclosure.py:100]: why you have added only these frus in fru_mapping list?? what about disk, controller, enclosure etc.

Comment (gargpalak048) [reply low-level/framework/platforms/realstor/realstor_enclosure.py:100]: so, FRUs through enclosure api '/show/frus'  are :- [POWER_SUPPLY, FAN MODULE, SIDEPLANE]
and in sspl we term as fan, psu, sideplane etc 
That's why this mapping is done.

cli api is not giving info related to 'disk', 'controller' and 'enclosure' FRUs
that I have kept in storage_fru_list>hot_swappable under global config as it is.
so mapping for these frus was not required.

File: low-level/actuators/impl/generic/realstor_encl.py [hunk 2]
```diff
@@ -93,16 +93,16 @@ def perform_request(self, jsonMsg):
...
             ctrl_type = ""
             if enclosure_request_data[-1] == "shutdown":
-                # "ENCL: enclosure:fru:controller:shutdown"
+                # "ENCL: enclosure:hw:controller:shutdown"
                 (request_type, _, component, component_type,
                     ctrl_action) = enclosure_request_data
             elif enclosure_request_data[-1] == "restart":
-                # "ENCL: enclosure:fru:controller:sc:restart"
-                # "ENCL: enclosure:fru:controller:mc:restart"
+                # "ENCL: enclosure:hw:controller:sc:restart"
```
Comment (cdeshmukh) [inline low-level/actuators/impl/generic/realstor_encl.py:100]: Make sure we communicate this to HA team as they plan to consume enclosure shutdown request for Feature F-8E, and also update the ticket with update request/response which introduced this.

File: low-level/framework/platforms/realstor/realstor_enclosure.py (hunk 1 above)
Comment (cdeshmukh) [reply low-level/framework/platforms/realstor/realstor_enclosure.py:100]: >and in sspl we term as fan, psu, sideplane etc

In SSPL also its seen as fan_modules, sideplane_expanders only

btw enclosure is not hot swappable. dont keep 'enclosure' resource in extended configurable fru list also

File: low-level/framework/platforms/realstor/realstor_enclosure.py [hunk 3]
```diff
@@ -483,5 +490,58 @@ def get_realstor_encl_data(self, fru: str):
...
+        # Read FRU data using /show/frus cli api.
+        fru_data = singleton_realstorencl.get_realstor_encl_data("frus")
+        for fru_dict in fru_data:
+            if "name" in fru_dict.keys():
+                self.fru_list.append(fru_dict["name"])
+        self.fru_list = list(set(self.fru_list))
+        try:
+            self.hot_swapped_frus = Conf.get(GLOBAL_CONF,
+                "storage_enclosure>storage_fru_list>hot_swappable",
+                ['disk', 'enclosure', 'controller'])
```
Comment (cdeshmukh) [inline low-level/framework/platforms/realstor/realstor_enclosure.py:517]: remove enclosure from fru list, dont keep in cold-swappable even by default

File: low-level/framework/platforms/realstor/realstor_enclosure.py (hunk 1 above)
Comment (gargpalak048) [reply low-level/framework/platforms/realstor/realstor_enclosure.py:100]: > In SSPL also its seen as fan_modules, sideplane_expanders only

we are using these terms while calling cli apis to get data.
But in sensor resource_type, it is as fan, sideplane etc.

Comment (gargpalak048) [reply low-level/framework/platforms/realstor/realstor_enclosure.py:100]: > btw enclosure is not hot swappable. dont keep 'enclosure' resource in extended configurable fru list also

ok removed from default list

**Review by shriya-deshmukh — APPROVED** [review]

**Review by sandeepsgt — APPROVED** [review]

**Review by jjxsg — APPROVED** [review]

File: low-level/framework/platforms/realstor/realstor_enclosure.py [hunk 4]
```diff
@@ -483,5 +490,58 @@ def get_realstor_encl_data(self, fru: str):
...
+            is_fru = False
+
+        fru_str = str(is_fru).lower()
+        if is_fru:
+            if fru in self.hot_swapped_frus:
+                fru_str = str(is_fru).lower() + ":" + "hot_swappable"
+            elif fru in self.cold_swapped_frus:
+                fru_str = str(is_fru).lower() + ":" + "cold_swappable"
+            else:
+                fru_str = str(is_fru).lower() + ":" + "unknown"
```
Comment (cdeshmukh) [inline low-level/framework/platforms/realstor/realstor_enclosure.py:542]: there not an unknown case, default to "cold_swappable"

File: low-level/actuators/impl/generic/realstor_encl.py (hunk 2 above)
Comment (gargpalak048) [reply low-level/actuators/impl/generic/realstor_encl.py:100]: > no unknown value we should keep, default should be made cold_swabble where no backend tells whether its hot swappable, including solution specific config.

So for fan-module, Power supply & sideplane FRUs, we should return default cold_swappable str?

Comment (gargpalak048) [reply low-level/actuators/impl/generic/realstor_encl.py:100]: done. Keeping the 'cold_swappable' as default unit.

>And please confirm we have tickets for HA & CSM to consume these changes, as applicable & capture in jira also

yes.. created tickets for same.

File: low-level/framework/utils/ipmi_client.py [hunk 5]
```diff
@@ -288,14 +288,11 @@ def is_fru(self, fru):
         if is_fru:
             if fru in self.hot_swapped_frus:
                 fru_str = str(is_fru).lower() + ":" + "hot_swappable"
-            elif fru in self.cold_swapped_frus:
-                fru_str = str(is_fru).lower() + ":" + "cold_swappable"
             else:
-                fru_str = str(is_fru).lower() + ":" + "unknown"
+                fru_str = str(is_fru).lower() + ":" + "cold_swappable"
```
Comment (gargpalak048) [inline low-level/framework/utils/ipmi_client.py:292]: addressing the same change for Server FRU representation as well in this PR only

File: low-level/framework/platforms/realstor/realstor_enclosure.py (hunk 4 above)
Comment (gargpalak048) [reply low-level/framework/platforms/realstor/realstor_enclosure.py]: done

**Review by thavanathan — APPROVED** [review]

**Review by cdeshmukh — APPROVED** [review]

**Comment by cdeshmukh** [comment]
test this please

**Review by mariyappanp — COMMENTED** [review]
Please complete PR checklist.

**Comment by gargpalak048** [comment]
> Please complete PR checklist.

done

**Review by mariyappanp — APPROVED** [review]

**Comment by cdeshmukh** [comment]
VM deployment passed
http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Cortx-Deployment/job/SSPL/527/

**Pull request merged**